TST use global_random_seed in sklearn/cluster/tests/test_bicluster.py

marenwestermann opened the pull request:
#### Reference Issues/PRs
towards #22827


#### What does this implement/fix? Explain your changes.
I used the global_random_seed fixture in some of the tests in the `tests/test_bicluster.py` module where I thought it's appropriate.

#### Any other comments?
`test_spectral_coclustering`: 
I tried using global_random_seed in this test which resulted in 6 test failures which I was able to fix by reducing `noise` from 0.5 to 0.1 (line 73). However, after the change the test finished in 17 seconds on my machine (compared to 6 seconds before). As this is quite long I didn't apply global_random_seed to this test. 

`test_spectral_biclustering`: 
When I used global_random_seed in this test, the test run time increased from 0.6 sec to 51 sec, so I left the test unchanged. 

`test_fit_best_piecewise`: 
Using global_random_seed in this test increased the test  run time from 0.08 sec to 2 sec. If I understand correctly, global_random_seed is not really needed in this test, so I left this test unchanged. 

`test_project_and_cluster`: 
Using global_random_seed here increased the test run time from from 0.08 sec to 1.51 sec. I think global_random_seed is also not really needed here, so I also left this test unchanged. 

`test_perfect_checkerboard`: 
Introducing global_random_seed to this test increased the test run time from 0.3 sec to 23.4 sec. From my understanding global_random_seed is also not of great importance here, I so didn't change this test.

Commit:
test_bicluster [all random seeds]
test_scale_normalize
test_bistochastic_normalize
test_log_normalize

marenwestermann commented:
ping @glemaitre

ogrisel commented:
> test_spectral_biclustering:
> When I used global_random_seed in this test, the test run time increased from 0.6 sec to 51 sec, so I left the test unchanged.

This is very surprising. I tried setting random seed to 42 instead of 0 for this test and the speed is still under a few seconds. Have you tried checking that your laptop was not overloaded with a background tasks (e.g. with `htop`) prior to launching this test?

glemaitre commented:
I think that @marenwestermann meant that running the test for all seeds is making the test much slower. Indeed this is not a problem because we are not going to run these tests most of the time but only target a single test. So we can add the fixture.

glemaitre commented:
@marenwestermann if you can confirm that the test become longer only when the environment variable `SKLEARN_TESTS_GLOBAL_RANDOM_SEED` is set, then I think that this is worth to parametrizing the tests that you discarded for the moment.

Commit:
test-bicluster [all random seeds]
test_spectral_coclustering
test_spectral_biclustering
test_fit_best_piecewise
test_project_and_cluster
test_perfect_checkerboard

marenwestermann commented:
> I think that @marenwestermann meant that running the test for all seeds is making the test much slower. Indeed this is not a problem because we are not going to run these tests most of the time but only target a single test. So we can add the fixture.

Yes, this is what I meant.

marenwestermann commented:
> @marenwestermann if you can confirm that the test become longer only when the environment variable `SKLEARN_TESTS_GLOBAL_RANDOM_SEED` is set, then I think that this is worth to parametrizing the tests that you discarded for the moment.

I can confirm this. I ran the tests again with and without `global_random_seed` and kept an eye on the CPU load which was low before running the tests. The run times for the tests where the same as reported above.

marenwestermann commented:
I added the `global_random_seed` fixture to the remaining tests.

glemaitre reviewed (approved):
So LGTM in this case. @ogrisel maybe you want to have a new look.

betatim commented:
I'm not Olivier but I took a moment to look at the comments and changes. They look good to me. All the instances that used a hard coded random seed now use `global_random_seed` and only one test needed modifying for it to pass with all 100 different seed values.

(Trying to get back into scikit-learn contributing, reviewing "good first issue" PRs seemed like a good way to do that. But it also feels a bit weird to turn up to a PR as a person unknown to most people here and making review comments :D)

jjerphan commented:
> (Trying to get back into scikit-learn contributing, reviewing "good first issue" PRs seemed like a good way to do that. But it also feels a bit weird to turn up to a PR as a person unknown to most people here and making review comments :D)

@betatim: honestly, feel free to review pull requests. We accept reviews from new people if those reviews make sense. In your case, you are also not new to the project nor the ecosystem and we would appreciate your contributions and help. :smiley: :handshake:

betatim commented:
Should I use the "review changes" feature of GitHub and leave a "approved" review that way? What is easiest for you?

jjerphan commented:
> Should I use the "review changes" feature of GitHub and leave a "approved" review that way? What is easiest for you?

Yes, it is useful for semantic and when querying GitHub's API.

jjerphan reviewed (approved):
LGTM. Thank you for this contribution and for your patience, @marenwestermann.

@betatim: I think you can pursue reviewing such PRs (to get started contributing back to scikit-learn if you are interested) and ping me in case of doubt.

jjerphan commented on sklearn/cluster/tests/test_bicluster.py:
```diff
@@ -165,44 +167,45 @@ def test_bistochastic_normalize():
+def test_perfect_checkerboard(global_random_seed):
     # XXX Previously failed on build bot (not reproducible)
-    model = SpectralBiclustering(3, svd_method="arpack", random_state=0)
+    random_state = global_random_seed
```
Oops, actually suggestions got lost in the tab-shuffle.

---
A single comment before merging: `global_random_seed` must in place of `random_state`.

Commit:
address comment

marenwestermann commented:
I updated the PR. Thank you everyone for reviewing! I much appreciate it! 🙂
